thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/DataAvailabilityTaskScheduler.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.pinot.thirdeye.formatter.DetectionConfigFormatter;
+import org.apache.pinot.thirdeye.rootcause.impl.MetricEntity;
+import org.apache.pinot.thirdeye.util.ThirdEyeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DataAvailabilityTaskScheduler implements Runnable {
+  public static final String LIX_FLAG_VALUE = "DATA_AVAILABILITY_SCHEDULE";
+  private static final Logger LOG = LoggerFactory.getLogger(DataAvailabilityTaskScheduler.class);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private ScheduledThreadPoolExecutor executorService;
+  private long delayInSec;
+  private int notRunThresholdInMin;
+  private Map<Long, Long> lastTaskCreateTimeMap;
+
+  public DataAvailabilityTaskScheduler(long delayInSec, int notRunThresholdInMin ) {
+    this.delayInSec = delayInSec;
+    this.notRunThresholdInMin = notRunThresholdInMin;
+    this.lastTaskCreateTimeMap = new HashMap<>();
+    this.executorService = new ScheduledThreadPoolExecutor(1, r -> {
+      Thread t = Executors.defaultThreadFactory().newThread(r);
+      t.setDaemon(true);
+      return t;
+    });
+  }
+
+  @Override
+  public void run() {
+    Multimap<Set<String>, DetectionConfigDTO> dataset2DetectionMap = ArrayListMultimap.create();
+    Map<String, Long> datasetRefreshTimeMap = new HashMap<>();
+    populateRequiredMetadata(dataset2DetectionMap, datasetRefreshTimeMap);
+    Map<Long, TaskDTO> runningDetection = retrieveRunningDetection();
+    for (Set<String> datasets : dataset2DetectionMap.keySet()) {
+      for (DetectionConfigDTO detectionConfig : dataset2DetectionMap.get(datasets)) {
+        long detectionConfigId = detectionConfig.getId();
+        long lastTimestamp = detectionConfig.getLastTimestamp();
+        boolean allUpdated = datasets.stream().map(d -> datasetRefreshTimeMap.get(d) >= lastTimestamp)
+            .reduce(true, (a, b) -> a && b);
+        if (allUpdated) {
+          if (!runningDetection.containsKey(detectionConfigId)) {
+            createDetectionTask(detectionConfig);
+          } else {
+            LOG.info("Skipping creating detection task for detection {} because task {} is not finished.",
+                detectionConfigId, runningDetection.get(detectionConfigId));
+          }
+        } else {
+          try {
+            if (!runningDetection.containsKey(detectionConfigId)) {
+              if (!lastTaskCreateTimeMap.containsKey(detectionConfigId)) loadLatestTaskCreateTime(detectionConfig);
+              long lastRunTime = lastTaskCreateTimeMap.get(detectionConfigId);
+              if (System.currentTimeMillis() - lastRunTime >= TimeUnit.MINUTES.toMillis(notRunThresholdInMin)) {
+                LOG.info("Scheduling a task for detection {}, since the create time of last task is {}",
+                    detectionConfigId, lastRunTime);
+                createDetectionTask(detectionConfig);
+              }
+            }
+          } catch (Exception e) {
```

Comment:
Suggest to refactor this code block. Split that into sub-methods which is easier to understand.